README.md
# @nkp/supports-color

[![npm version](https://badge.fury.io/js/%40nkp%2Fsupports-color.svg)](https://www.npmjs.com/package/@nkp/supports-color)
[![deploy status](https://github.com/nickkelly1/supports-color/actions/workflows/release.yml/badge.svg)](https://github.com/nickkelly1/supports-color/actions/workflows/release.yml)
[![known vulnerabilities](https://snyk.io/test/github/nickkelly1/supports-color/badge.svg)](https://snyk.io/test/github/nickkelly1/supports-color)

Detect whether a terminal supports color.

Fork of [`supports-color`](https://github.com/chalk/supports-color) with support for both ESM and CommonJS.

## Table of contents

- [Exports](#exports)
- [Fork](#fork)
- [Installation](#installation)
  - [npm](#npm)
  - [yarn](#yarn)
  - [pnpm](#pnpm)
- [Usage](#usage)
- [publishing](#publishing)

## Exports

`@nkp/supports-color` exports both CommonJS and ES modules.

## Fork

`@nkp/supports-color` is a fork of [`supports-color`](https://github.com/chalk/supports-color) that exports both ESM and CommonJS.

In 2021 [Sindre Sorhus](https://gist.github.com/sindresorhus), prolific creator and maintainer of many popular [`npm packages`](https://www.npmjs.com/~sindresorhus), [including some with over 150 million weekly monthly downloads](https://www.npmjs.com/package/chalk), pushed a new major version of many of his packages removing CommonJS support and making them ["pure ESM"](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c). This was a major breaking change. Sindre's universally popular packages were now incompatible with a large parts of the NodeJS ecosystem.

Sindre's packages could have been kept backwards compatible by producing both ESM and CJS exports, however [Sindre believed he could push the NodeJS community to adopt ESM sooner by introducing breaking changes](https://blog.sindresorhus.com/get-ready-for-esm-aa53530b3f77).

> Personally, I plan to (migrate to Pure ESM *instead* of provide backwards compatibility) as I think it’s better to rip off the bandaid and push the ecosystem forward.

[For many, the breaking change's haven't gone over well](https://gist.github.com/joepie91/bca2fda868c1e8b2c2caf76af7dfcad3). Many project's now find themselves spending hours, days or even weeks trying to placate these pure EMS packages only to find too many incompatibilities and revert back.

Hence this fork. `@nkp/supports-color` is intended to be an functionally equivalent clone of `styles-ansi` but with both ESM and CJS exports.

## Installation

### npm

```sh
npm install @nkp/supports-color
```

### yarn

```sh
yarn add @nkp/supports-color
```

### pnpm

```sh
pnpm add @nkp/supports-color
```

## Publishing

To a release a new version:

1. Update the version number in package.json
2. Push the new version to the `master` branch on GitHub
3. Create a `new release` on GitHub for the latest version

This will trigger a GitHub action that tests and publishes the npm package.


changelog.md
# Changelog

## 0.0.3 - 2022-02-11

### Changed

- Updated dependencies
- Updated readme.md
- Updated build process
- Updated package.json exports

## 0.0.2 - 2022-01-23

### Changed

- Updated package.json exports


## 0.0.1 - 2022-01-23

### Added

- Forked & rewrote


tests/test.ts
import process from 'process';
import os from 'os';
import tty from 'tty';
import { ColorSupportObject, SupportsColor } from '../src/types';
import { createNodeSupportsColor } from '../src/node.create';

const currentNodeVersion = process.versions.node;

describe('@nkp/supports-colors', () => {
  beforeEach(() => {
    Object.defineProperty(process, 'platform', {
      value: 'linux',
    });

    Object.defineProperty(process.versions, 'node', {
      value: currentNodeVersion,
    });

    process.stdout.isTTY = true;
    process.argv = [];
    process.env = {};
    tty.isatty = () => true;
  });

  // FIXME
  it('return true if `FORCE_COLOR` is in env', async () => {
    process.stdout.isTTY = false;
    process.env['FORCE_COLOR'] = true as any;
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(1);
  });

  it('return true if `FORCE_COLOR` is in env, but honor 256', async () => {
    process.argv = ['--color=256',];
    process.env['FORCE_COLOR'] = true as any;
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
  });

  it('return true if `FORCE_COLOR` is in env, but honor 256 #2', async () => {
    process.argv = ['--color=256',];
    process.env['FORCE_COLOR'] = '1';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
  });

  it('return false if `FORCE_COLOR` is in env and is 0', async () => {
    process.env['FORCE_COLOR'] = '0';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('do not cache `FORCE_COLOR`', async () => {
    process.env['FORCE_COLOR'] = '0';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
    process.env['FORCE_COLOR'] = '1';
    const updatedStdOut = result({ isTTY: tty.isatty(1), });
    expect(updatedStdOut).toBeTruthy();
  });

  it('return false if not TTY', async () => {
    process.stdout.isTTY = false;
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return false if --no-color flag is used', async () => {
    process.env = { TERM: 'xterm-256color', };
    process.argv = ['--no-color',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return false if --no-colors flag is used', async () => {
    process.env = {TERM: 'xterm-256color',};
    process.argv = ['--no-colors',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return true if --color flag is used', async () => {
    process.argv = ['--color',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('return true if --colors flag is used', async () => {
    process.argv = ['--colors',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('return true if `COLORTERM` is in env', async () => {
    process.env['COLORTERM'] = true as any;
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('support `--color=true` flag', async () => {
    process.argv = ['--color=true',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('support `--color=always` flag', async () => {
    process.argv = ['--color=always',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('support `--color=false` flag', async () => {
    process.env = {TERM: 'xterm-256color',};
    process.argv = ['--color=false',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('support `--color=256` flag', async () => {
    process.argv = ['--color=256',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('level should be 2 if `--color=256` flag is used', async () => {
    process.argv = ['--color=256',];
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
    expect((result.stdout as ColorSupportObject).has256).toBeTruthy();
  });

  it('support `--color=16m` flag', async () => {
    process.argv = ['--color=16m',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('support `--color=full` flag', async () => {
    process.argv = ['--color=full',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('support `--color=truecolor` flag', async () => {
    process.argv = ['--color=truecolor',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('level should be 3 if `--color=16m` flag is used', async () => {
    process.argv = ['--color=16m',];
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(3);
    expect((result.stdout as ColorSupportObject).has256).toBeTruthy();
    expect((result.stdout as ColorSupportObject).has16m).toBeTruthy();
  });

  it('ignore post-terminator flags', async () => {
    process.argv = ['--color', '--', '--no-color',];
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('allow tests of the properties on false', async () => {
    process.env = {TERM: 'xterm-256color',};
    process.argv = ['--no-color',];
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).hasBasic).toBe(undefined);
    expect((result.stdout as ColorSupportObject).has256).toBe(undefined);
    expect((result.stdout as ColorSupportObject).has16m).toBe(undefined);
    expect((result.stdout as ColorSupportObject).level).toBeUndefined();
  });

  it('return false if `CI` is in env', async () => {
    process.env['CI'] = 'AppVeyor';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return true if `TRAVIS` is in env', async () => {
    process.env = { CI: 'Travis', TRAVIS: '1', };
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('return true if `CIRCLECI` is in env', async () => {
    process.env = { CI: true as any, CIRCLECI: true as any, };
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('return true if `APPVEYOR` is in env', async () => {
    process.env = { CI: true as any, APPVEYOR: true as any, };
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('return true if `GITLAB_CI` is in env', async () => {
    process.env = { CI: true as any, GITLAB_CI: true as any,};
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('return true if `BUILDKITE` is in env', async () => {
    process.env = { CI: true as any, BUILDKITE: true as any, };
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('return true if `DRONE` is in env', async () => {
    process.env = { CI: true as any, DRONE: true as any, };
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
  });

  it('return true if Codeship is in env', async () => {
    process.env = { CI: true as any, CI_NAME: 'codeship', };
    const result = createNodeSupportsColor();
    expect(result).toBeTruthy();
  });

  it('return false if `TEAMCITY_VERSION` is in env and is < 9.1', async () => {
    process.env['TEAMCITY_VERSION'] = '9.0.5 (build 32523)';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return level 1 if `TEAMCITY_VERSION` is in env and is >= 9.1', async () => {
    process.env['TEAMCITY_VERSION'] = '9.1.0 (build 32523)';
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(1);
  });

  it('support rxvt', async () => {
    process.env = { TERM: 'rxvt', };
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(1);
  });

  it('prefer level 2/xterm over COLORTERM', async () => {
    process.env = { COLORTERM: '1', TERM: 'xterm-256color', };
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
  });

  it('support screen-256color', async () => {
    process.env = { TERM: 'screen-256color', };
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
  });

  it('support putty-256color', async () => {
    process.env = { TERM: 'putty-256color', };
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
  });

  it('level should be 3 when using iTerm 3.0', async () => {
    Object.defineProperty(process, 'platform', {
      value: 'darwin',
    });
    process.env = {
      TERM_PROGRAM: 'iTerm.app',
      TERM_PROGRAM_VERSION: '3.0.10',
    };
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(3);
  });

  it('level should be 2 when using iTerm 2.9', async () => {
    Object.defineProperty(process, 'platform', {
      value: 'darwin',
    });
    process.env = {
      TERM_PROGRAM: 'iTerm.app',
      TERM_PROGRAM_VERSION: '2.9.3',
    };
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
  });

  it('return level 1 if on Windows earlier than 10 build 10586', async () => {
    Object.defineProperty(process, 'platform', {
      value: 'win32',
    });
    Object.defineProperty(process.versions, 'node', {
      value: '8.0.0',
    });
    os.release = () => '10.0.10240';
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(1);
  });

  it('return level 2 if on Windows 10 build 10586 or later', async () => {
    Object.defineProperty(process, 'platform', {
      value: 'win32',
    });
    Object.defineProperty(process.versions, 'node', {
      value: '8.0.0',
    });
    os.release = () => '10.0.10586';
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
  });

  it('return level 3 if on Windows 10 build 14931 or later', async () => {
    Object.defineProperty(process, 'platform', {
      value: 'win32',
    });
    Object.defineProperty(process.versions, 'node', {
      value: '8.0.0',
    });
    os.release = () => '10.0.14931';
    const result = createNodeSupportsColor();
    expect((result.stdout as ColorSupportObject).level).toBe(3);
  });

  it('return level 2 when FORCE_COLOR is set when not TTY in xterm256', async () => {
    process.stdout.isTTY = false;
    process.env['FORCE_COLOR'] = true as any;
    process.env['TERM'] = 'xterm-256color';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(2);
  });

  it('supports setting a color level using FORCE_COLOR', async () => {
    let result: SupportsColor;
    process.env['FORCE_COLOR'] = '1';
    result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(1);

    process.env['FORCE_COLOR'] = '2';
    result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(2);

    process.env['FORCE_COLOR'] = '3';
    result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(3);

    process.env['FORCE_COLOR'] = '0';
    result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('FORCE_COLOR maxes out at a value of 3', async () => {
    process.env['FORCE_COLOR'] = '4';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(3);
  });

  it('FORCE_COLOR works when set via command line (all values are strings)', async () => {
    let result: SupportsColor;
    process.env['FORCE_COLOR'] = 'true';
    result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(1);

    process.stdout.isTTY = false;
    process.env['FORCE_COLOR'] = 'true';
    process.env['TERM'] = 'xterm-256color';
    result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(2);

    process.env['FORCE_COLOR'] = 'false';
    result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return false when `TERM` is set to dumb', async () => {
    process.env['TERM'] = 'dumb';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return false when `TERM` is set to dumb when `TERM_PROGRAM` is set', async () => {
    process.env['TERM'] = 'dumb';
    process.env['TERM_PROGRAM'] = 'Apple_Terminal';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return false when `TERM` is set to dumb when run on Windows', async () => {
    Object.defineProperty(process, 'platform', {
      value: 'win32',
    });
    Object.defineProperty(process.versions, 'node', {
      value: '10.13.0',
    });
    os.release = () => '10.0.14931';
    process.env['TERM'] = 'dumb';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeFalsy();
  });

  it('return level 1 when `TERM` is set to dumb when `FORCE_COLOR` is set', async () => {
    process.env['FORCE_COLOR'] = '1';
    process.env['TERM'] = 'dumb';
    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(1);
  });

  it('ignore flags when sniffFlags=false', async () => {
    process.argv = ['--color=256',];
    process.env['TERM'] = 'dumb';

    const result = createNodeSupportsColor();
    expect(result.stdout).toBeTruthy();
    expect((result.stdout as ColorSupportObject).level).toBe(2);

    const sniffResult = result(process.stdout, {sniffFlags: true,});
    expect(sniffResult).toBeTruthy();
    expect((sniffResult as ColorSupportObject).level).toBe(2);

    const nosniffResult = result(process.stdout, {sniffFlags: false,});
    expect(nosniffResult).toBeFalsy();
  });
});

src/node.factory.ts
import tty from 'tty';
import { HasTTY, SupportsColor } from './types';
import process from 'process';
import os from 'os';
import { ColorSupport } from './types';

function colorSupportFactoryNode(): SupportsColor {
  // From: https://github.com/sindresorhus/has-flag/blob/main/index.js
  function hasFlag(flag: string, argv = process.argv) {
    const prefix = flag.startsWith('-') ? '' : (flag.length === 1 ? '-' : '--');
    const position = argv.indexOf(prefix + flag);
    const terminatorPosition = argv.indexOf('--');
    return position !== -1 && (terminatorPosition === -1 || position < terminatorPosition);
  }

  const { env, } = process;

  let flagForceColor: number;
  if (
    hasFlag('no-color')
    || hasFlag('no-colors')
    || hasFlag('color=false')
    || hasFlag('color=never')
  ) {
    flagForceColor = 0;
  } else if (
    hasFlag('color')
    || hasFlag('colors')
    || hasFlag('color=true')
    || hasFlag('color=always')
  ) {
    flagForceColor = 1;
  }

  function isTrue(value: null | undefined | boolean | string) {
    if (value == null) return false;
    if (typeof value === 'boolean') return value;
    if (value.trim().toLowerCase() === 'true') return true;
    return false;
  }

  function isFalse(value: null | undefined | boolean | string) {
    if (value == null) return false;
    if (typeof value === 'boolean') return !value;
    if (value.trim().toLowerCase() === 'false') return true;
    return false;
  }

  function envForceColor(): number | void {
    if ('FORCE_COLOR' in env) {
      if (isTrue(env['FORCE_COLOR'])) {
        return 1;
      }

      if (isFalse(env['FORCE_COLOR'])) {
        return 0;
      }

      return env['FORCE_COLOR']!.length === 0 ? 1 : Math.min(Number.parseInt(env['FORCE_COLOR']!, 10), 3);
    }
  }

  function translateLevel(level: number): ColorSupport {
    if (level === 0) {
      return false;
    }

    return {
      level,
      hasBasic: true,
      has256: level >= 2,
      has16m: level >= 3,
    };
  }

  function _supportsColor(
    haveStream: HasTTY,
    { streamIsTTY, sniffFlags = true, }: { streamIsTTY?: boolean, sniffFlags?: boolean, } = {}
  ) {
    const noFlagForceColor = envForceColor();
    if (noFlagForceColor !== undefined) {
      flagForceColor = noFlagForceColor;
    }

    const forceColor = sniffFlags ? flagForceColor : noFlagForceColor;

    if (forceColor === 0) {
      return 0;
    }

    if (sniffFlags) {
      if (hasFlag('color=16m')
        || hasFlag('color=full')
        || hasFlag('color=truecolor')) {
        return 3;
      }

      if (hasFlag('color=256')) {
        return 2;
      }
    }

    if (haveStream && !streamIsTTY && forceColor === undefined) {
      return 0;
    }

    const min = forceColor || 0;

    if (env['TERM'] === 'dumb') {
      return min;
    }

    if (process.platform === 'win32') {
      // Windows 10 build 10586 is the first Windows release that supports 256 colors.
      // Windows 10 build 14931 is the first release that supports 16m/TrueColor.
      const osRelease = os.release().split('.');
      if (
        Number(osRelease[0]) >= 10
        && Number(osRelease[2]) >= 10_586
      ) {
        return Number(osRelease[2]) >= 14_931 ? 3 : 2;
      }

      return 1;
    }

    if ('CI' in env) {
      if (['TRAVIS', 'CIRCLECI', 'APPVEYOR', 'GITLAB_CI', 'GITHUB_ACTIONS', 'BUILDKITE', 'DRONE',].some(sign => sign in env) || env['CI_NAME'] === 'codeship') {
        return 1;
      }

      return min;
    }

    if ('TEAMCITY_VERSION' in env) {
      return /^(9\.(0*[1-9]\d*)\.|\d{2,}\.)/.test(env['TEAMCITY_VERSION']!) ? 1 : 0;
    }

    // Check for Azure DevOps pipelines
    if ('TF_BUILD' in env && 'AGENT_NAME' in env) {
      return 1;
    }

    if (env['COLORTERM'] === 'truecolor') {
      return 3;
    }

    if ('TERM_PROGRAM' in env) {
      const version = Number.parseInt((env['TERM_PROGRAM_VERSION']! || '').split('.')[0]!, 10);

      switch (env['TERM_PROGRAM']!) {
        case 'iTerm.app':
          return version >= 3 ? 3 : 2;
        case 'Apple_Terminal':
          return 2;
        // No default
      }
    }

    if (/-256(color)?$/i.test(env['TERM']!)) {
      return 2;
    }

    if (/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env['TERM']!)) {
      return 1;
    }

    if ('COLORTERM' in env) {
      return 1;
    }

    return min;
  }

  const supportsColor: SupportsColor = function supportsColorNode(
    stream: HasTTY,
    options = {}
  ): ColorSupport {
    const level = _supportsColor(stream, {
      streamIsTTY: stream && (stream as tty.ReadStream).isTTY,
      ...options,
    });

    return translateLevel(level);
  } as SupportsColor;


  supportsColor.stdout = supportsColor({ isTTY: tty.isatty(1), });
  supportsColor.stderr = supportsColor({ isTTY: tty.isatty(2), });

  return supportsColor;
}

export default colorSupportFactoryNode;

config/rollup.config.js
import peerDepsExternal from 'rollup-plugin-peer-deps-external';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import typescript from 'rollup-plugin-typescript2';
import json from '@rollup/plugin-json';
// import packageJson from '../package.json';

/** @typedef {import('rollup').RollupOptions} RollupOptions */

/** @type {RollupOptions} */
const config = {
  // input: 'src/index.ts',
  input: 'src',
  // input: { },
  external: [/node_modules/, /^node:/,],
  output: [
    {
      // file: 'dist/cjs/index.js',
      dir: 'dist/cjs',
      interop: 'auto',
      format: 'cjs',
      sourcemap: false,
    },
    // {
    //   file: 'dist/es/index.js',
    //   interop: 'auto',
    //   format: 'es',
    //   exports: 'named',
    //   sourcemap: false,
    // },
  ],
  plugins: [
    // @ts-ignore peerDepxExternal is typed wrong
    /** @type {Plugin} */ peerDepsExternal(),
    resolve(),
    commonjs(),
    typescript({ tsconfig: 'config/tsconfig.build.json', }),
    json(),
  ],
};

/** @type {RollupOptions[]} */
export default [config,];


src/node.create.ts
import tty from 'tty';
import { HasTTY, SupportsColor } from './types';
import process from 'process';
import os from 'os';
import { ColorSupport } from './types';

// require factory function for testing
export function createNodeSupportsColor(): SupportsColor {
  // From: https://github.com/sindresorhus/has-flag/blob/main/index.js
  function hasFlag(flag: string, argv = process.argv) {
    const prefix = flag.startsWith('-') ? '' : (flag.length === 1 ? '-' : '--');
    const position = argv.indexOf(prefix + flag);
    const terminatorPosition = argv.indexOf('--');
    return position !== -1 && (terminatorPosition === -1 || position < terminatorPosition);
  }

  const { env, } = process;

  let flagForceColor: number;
  if (
    hasFlag('no-color')
    || hasFlag('no-colors')
    || hasFlag('color=false')
    || hasFlag('color=never')
  ) {
    flagForceColor = 0;
  } else if (
    hasFlag('color')
    || hasFlag('colors')
    || hasFlag('color=true')
    || hasFlag('color=always')
  ) {
    flagForceColor = 1;
  }

  function isTrue(value: null | undefined | boolean | string) {
    if (value == null) return false;
    if (typeof value === 'boolean') return value;
    if (value.trim().toLowerCase() === 'true') return true;
    return false;
  }

  function isFalse(value: null | undefined | boolean | string) {
    if (value == null) return false;
    if (typeof value === 'boolean') return !value;
    if (value.trim().toLowerCase() === 'false') return true;
    return false;
  }

  function envForceColor(): number | void {
    if ('FORCE_COLOR' in env) {
      if (isTrue(env['FORCE_COLOR'])) {
        return 1;
      }

      if (isFalse(env['FORCE_COLOR'])) {
        return 0;
      }

      return env['FORCE_COLOR']!.length === 0 ? 1 : Math.min(Number.parseInt(env['FORCE_COLOR']!, 10), 3);
    }
  }

  function translateLevel(level: number): ColorSupport {
    if (level === 0) {
      return false;
    }

    return {
      level,
      hasBasic: true,
      has256: level >= 2,
      has16m: level >= 3,
    };
  }

  function _supportsColor(
    haveStream: HasTTY,
    { streamIsTTY, sniffFlags = true, }: { streamIsTTY?: boolean, sniffFlags?: boolean, } = {}
  ) {
    const noFlagForceColor = envForceColor();
    if (noFlagForceColor !== undefined) {
      flagForceColor = noFlagForceColor;
    }

    const forceColor = sniffFlags ? flagForceColor : noFlagForceColor;

    if (forceColor === 0) {
      return 0;
    }

    if (sniffFlags) {
      if (hasFlag('color=16m')
        || hasFlag('color=full')
        || hasFlag('color=truecolor')) {
        return 3;
      }

      if (hasFlag('color=256')) {
        return 2;
      }
    }

    if (haveStream && !streamIsTTY && forceColor === undefined) {
      return 0;
    }

    const min = forceColor || 0;

    if (env['TERM'] === 'dumb') {
      return min;
    }

    if (process.platform === 'win32') {
      // Windows 10 build 10586 is the first Windows release that supports 256 colors.
      // Windows 10 build 14931 is the first release that supports 16m/TrueColor.
      const osRelease = os.release().split('.');
      if (
        Number(osRelease[0]) >= 10
        && Number(osRelease[2]) >= 10_586
      ) {
        return Number(osRelease[2]) >= 14_931 ? 3 : 2;
      }

      return 1;
    }

    if ('CI' in env) {
      if (['TRAVIS', 'CIRCLECI', 'APPVEYOR', 'GITLAB_CI', 'GITHUB_ACTIONS', 'BUILDKITE', 'DRONE',].some(sign => sign in env) || env['CI_NAME'] === 'codeship') {
        return 1;
      }

      return min;
    }

    if ('TEAMCITY_VERSION' in env) {
      return /^(9\.(0*[1-9]\d*)\.|\d{2,}\.)/.test(env['TEAMCITY_VERSION']!) ? 1 : 0;
    }

    // Check for Azure DevOps pipelines
    if ('TF_BUILD' in env && 'AGENT_NAME' in env) {
      return 1;
    }

    if (env['COLORTERM'] === 'truecolor') {
      return 3;
    }

    if ('TERM_PROGRAM' in env) {
      const version = Number.parseInt((env['TERM_PROGRAM_VERSION']! || '').split('.')[0]!, 10);

      switch (env['TERM_PROGRAM']!) {
        case 'iTerm.app':
          return version >= 3 ? 3 : 2;
        case 'Apple_Terminal':
          return 2;
        // No default
      }
    }

    if (/-256(color)?$/i.test(env['TERM']!)) {
      return 2;
    }

    if (/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env['TERM']!)) {
      return 1;
    }

    if ('COLORTERM' in env) {
      return 1;
    }

    return min;
  }

  const supportsColor: SupportsColor = function supportsColorNode(
    stream: HasTTY,
    options = {}
  ): ColorSupport {
    const level = _supportsColor(stream, {
      streamIsTTY: stream && (stream as tty.ReadStream).isTTY,
      ...options,
    });

    return translateLevel(level);
  } as SupportsColor;

  supportsColor.stdout = supportsColor({ isTTY: tty.isatty(1), });
  supportsColor.stderr = supportsColor({ isTTY: tty.isatty(2), });

  return supportsColor;
}
